### PR TITLE
 #167765398 User is able to get a Profile upon registration

### PR DESCRIPTION
**Pull request by voke1**
## What does this PR do?
- A User is able to update profile info during registration
- A User is able to get profile info upon registration

## Description of Task to be completed?
- Write controller function in (src/controllers/userController.js) that would use the user models to update user.
- Write unit tests
- Document using swagger

## How should this be manually tested?
- Run `npm install`
- Run ``npm run test`` to test the branch

### To edit profile
 send a PATCH request containing the relevant user fields to `/api/v1/auth/user/update-profile`

```
- Sample request body
{
	"firstName":"john",
	"lastName": "doe",
	"password": "password",
	"phoneNumber": "0901234567",
        "gender": "male",
        "preferredLanguage": "English",
        "department": "IT",
        "residenceAddress": "25, Jeremiah Ugwu, Lekki, Lagos.,
        "birthDate": "23-10-1898",
}
```
### To get user profile
- send a GET request containing the relevant user fields to ` /api/v1/auth/user/get-profile`

## What are the relevant pivotal tracker stories?
- #167765398

## Implementation plan

https://docs.google.com/document/d/1vCirHsSGDPKfplV_08ARIszkDym3oHZTlIUd4qzCgyk/edit?usp=sharing

image

image

### CODE REVIEW

File: src/controllers/UserController.js [hunk 1]
```diff
@@ -163,4 +162,28 @@ export default class UserController {
...
+ */
+  static async updateUserInfo(req, res) {
+    try {
+      const userData = { ...req.body };
+      const { user } = req;
+      userData.password = Utils.hashPassword(userData.password);
+      const data = await UserServices.updateUserInfoById({ ...userData }, user.email);
+      const userInfo = data[1].dataValues;
+      if (userInfo) {
+        delete userInfo.password;
```
Comment (victor-nach) [inline src/controllers/UserController.js]: I think it's better for you to do this in the user services file, delete the password from there first, also you can destructure properly there so that, what your returning to the controller is the userInfo ready to be returned to the user

File: src/controllers/UserController.js [hunk 2]
```diff
@@ -163,4 +162,28 @@ export default class UserController {
...
+        return resSuccess(res, 201, userInfo);
+      }
+      return resError(res, 404, 'User does not exist');
```
Comment (victor-nach) [inline src/controllers/UserController.js]: I think it's also better for you to check for this in the middleware, so that by the time it get's here we are sure the user already exists, if for any other reason the user services fail, the error would be caught in the catch block.

File: src/models/user.js [hunk 3]
```diff
@@ -17,20 +46,28 @@ export default (sequelize, DataTypes) => {
...
+      defaultValue: false
+    },
     phoneNumber: {
       allowNull: true,
       type: DataTypes.STRING
     }
   });
+
+  User.associate = (models) => {
+    User.belongsToMany(models.Role, {
```
Comment (victor-nach) [inline src/models/user.js]: Is the PR where roles are meant to implemented, I thought that was ugo's branch?

File: src/routes/userRoutes.js [hunk 4]
```diff
@@ -14,10 +14,12 @@ const {
   getGoogleAccountFromCode,
   getLinkedinUrl,
   getLinkedinAccountFromCode,
+  updateUserInfo
 } = UserController;
 
+
 const { checkUserExists, checkUserExistBeforeLogin } = UserMiddlewares;
-const { emailValidation, userLoginValidation } = userValidations;
+const { emailValidation, userValidation, userLoginValidation } = userValidations;
```
Comment (victor-nach) [inline src/routes/userRoutes.js]: I think you should change this name to updateInfoValidation or something more descriptive.

File: src/routes/userRoutes.js [hunk 5]
```diff
@@ -27,5 +29,7 @@ userRoutes.get('/user/google/signin', getGoogleUrl);
 userRoutes.get('/google/callback', getGoogleAccountFromCode);
 userRoutes.get('/user/linkedin/signin', getLinkedinUrl);
 userRoutes.get('/linkedin/callback', getLinkedinAccountFromCode);
+userRoutes.patch('/user/update-profile', userValidation, validationHandler, verifyToken, updateUserInfo);
```
Comment (victor-nach) [inline src/routes/userRoutes.js]: I should you should check for token first, before doing anything else.

Comment (voke1) [reply src/routes/userRoutes.js]: noted

File: src/routes/userRoutes.js (hunk 4 above)
Comment (voke1) [reply src/routes/userRoutes.js]: Noted

File: src/controllers/UserController.js (hunk 1 above)
Comment (voke1) [reply src/controllers/UserController.js]: Noted

File: src/models/user.js (hunk 3 above)
Comment (voke1) [reply src/models/user.js]: There has to be an associate function in the model file, `users` and `role`. Ugo can however, update.

File: src/controllers/UserController.js (hunk 1 above)
Comment (voke1) [reply src/controllers/UserController.js]: Good suggestion. done

File: src/middlewares/userMiddlewares.js [hunk 6]
```diff
@@ -47,4 +47,23 @@ export default class UserMiddlewares {
...
+
+  /**
+ * @name confirmUserExists
+ * @description Checks if a user exists in the database
+ * @param {object} req The request object
+ * @param {object} res The response object
+ * @param {object} next The response object
+ * @returns {object} The API response or next()
+ */
+  static async confirmUserExists(req, res, next) {
```
Comment (victor-nach) [inline src/middlewares/userMiddlewares.js:59]: This function is a sort of a duplicate function, adding this would mean we now have three different middlewares that call the same user service. I've looked at the code, I think this may be okay for now, because trying to combine it may make you touch other aspects of the code that this PR doesn't cover. we may have to open a branch that refactors this.

**Review by victor-nach — APPROVED** [review]
I've tested this locally, it works on postman, and all the tests pass. Nice job voke.

File: src/controllers/UserController.js (hunk 2 above)
Comment (voke1) [reply src/controllers/UserController.js]: Done

File: src/routes/userRoutes.js (hunk 4 above)
Comment (voke1) [reply src/routes/userRoutes.js]: Done

File: README.md [hunk 7]
```diff
@@ -1,6 +1,6 @@
...
 
-[![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)  [![Build Status](https://travis-ci.com/andela/phoenix-bn-backend.svg?branch=develop)](https://travis-ci.com/andela/phoenix-bn-backend) [![Maintainability](https://api.codeclimate.com/v1/badges/28351a9e117de0ecbba0/maintainability)](https://codeclimate.com/github/andela/phoenix-bn-backend/maintainability) [![Coverage Status](https://coveralls.io/repos/github/andela/phoenix-bn-backend/badge.svg?branch=develop)](https://coveralls.io/github/andela/phoenix-bn-backend?branch=develop)
+[![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)  [![Build Status](https://travis-ci.org/andela/phoenix-bn-backend.svg?branch=develop)](https://travis-ci.org/andela/phoenix-bn-backend) [![Maintainability](https://api.codeclimate.com/v1/badges/28351a9e117de0ecbba0/maintainability)](https://codeclimate.com/github/andela/phoenix-bn-backend/maintainability) [![Test Coverage](https://api.codeclimate.com/v1/badges/28351a9e117de0ecbba0/test_coverage)](https://codeclimate.com/github/andela/phoenix-bn-backend/test_coverage)
```
Comment (anneKay) [inline README.md:4]: Why are you making this change here?

File: public/api-docs/swaggerDoc.json [hunk 8]
```diff
@@ -447,9 +525,37 @@
...
+                    "type": "string",
+                    "description": "User role"
+                },
+                "preferredLanguage": {
+                    "type": "string",
+                    "description": "User preferred language"
+                },
+                "preferredCurrency": {
                     "type": "string",
-                    "description": "Address of WeTravel User"
```
Comment (anneKay) [inline public/api-docs/swaggerDoc.json:452]: Please revert this change, you shouldn't be removing other people's changes when adding yours unless that is your intention

File: public/api-docs/swaggerDoc.json [hunk 9]
```diff
@@ -447,9 +525,37 @@
...
+                "gender": {
+                    "type": "string",
+                    "description": "User gender"
+                },
+                "phoneNumber": {
+                    "type": "integer",
+                    "description": "Phone number of user"
+                },
+                "role": {
+                    "type": "string",
```
Comment (anneKay) [inline public/api-docs/swaggerDoc.json]: This is not a string, this should be an integer

File: src/middlewares/userMiddlewares.js (hunk 6 above)
Comment (anneKay) [reply src/middlewares/userMiddlewares.js:59]: There is a `checkUserExistBeforeLogin` function in the codebase, reuse that, but rename it to `checkIfUserExist`, remember to also rename this in this login route where it was used

File: README.md (hunk 7 above)
Comment (voke1) [reply README.md:4]: Okay. I only made change to the Travis CI badge and this is so we are able to view builds on Travis-CI by clicking on the badge. The repo is currently not activated on `travis-ci.com` which is currently present in the badge, instead it is on `travis-ci.org`.

File: public/api-docs/swaggerDoc.json (hunk 9 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json]: Noted

File: public/api-docs/swaggerDoc.json (hunk 8 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json]: I actually replaced it with ```  "residencAddress": {
                    "type": "string",
                    "description": "residence address of the user"
                }, ``` I'll revert however.

File: public/api-docs/swaggerDoc.json (hunk 9 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json]: Noted

File: public/api-docs/swaggerDoc.json (hunk 8 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json:452]: I actually replaced it with ```  "residencAddress": {
                    "type": "string",
                    "description": "residence address of the user"
                }, ``` I'll revert however.

File: README.md (hunk 7 above)
Comment (anneKay) [reply README.md:4]: I can see that you replaced the coveralls badge with codeclimate badge

File: public/api-docs/swaggerDoc.json (hunk 8 above)
Comment (anneKay) [reply public/api-docs/swaggerDoc.json:452]: Okay then, that's fine

File: src/controllers/UserController.js [hunk 10]
```diff
@@ -59,7 +58,7 @@ export default class UserController {
         return resSuccess(res, 200, user);
       }
       if (!isCorrectPassword) {
-        return resError(res, 401, 'Inavalid email/password.');
+        return resError(res, 401, 'Invalid email/password.');
```
Comment (anneKay) [inline src/controllers/UserController.js:61]: Thanks for spotting that

File: README.md (hunk 7 above)
Comment (voke1) [reply README.md:4]: That's my error. I've reverted that. Thanks

File: public/api-docs/swaggerDoc.json (hunk 9 above)
Comment (voke1) [reply public/api-docs/swaggerDoc.json]: Done

File: README.md (hunk 7 above)
Comment (voke1) [reply README.md]: The `checkifUserExistBeforeLogin` checks if user exist during login and gets user email from req.body.
However, the `confirmUserExists` checks if user is valid during registration. It gets user email  from token sent to the user. As user should not be allowed to update his email during registration.

The two middleware are not exactly the same. Please advise.

File: src/validation/userValidation.js [hunk 11]
```diff
@@ -1,19 +1,26 @@
...
     .withMessage('First name field cannot be empty.'),
   check('lastName').trim().not().isEmpty()
     .withMessage('last name field cannot be empty.'),
-  check('email').trim().not().isEmpty()
-    .withMessage('Email field cannot be empty'),
-  check('email').isEmail().withMessage('Enter valid email address.'),
-  check('email').normalizeEmail(),
   check('password').trim().not().isEmpty()
     .withMessage('Please password is required'),
   check('password').isLength({ min: 8 }).withMessage('Password should be atleast 8 characters'),
```
Comment (anneKay) [inline src/validation/userValidation.js]: The user will not be changing their password on their profile. There should be a reset password feature for that

**Review by anneKay — CHANGES_REQUESTED** [review]
Please have a get `/profile` route which should call a getProfile method and return a user's profile. Then also have an edit profile route which would call another method and enable a user edit their profile

**Comment by voke1** [comment]
> Please have a get `/profile` route which should call a getProfile method and return a user's profile. Then also have an edit profile route which would call another method and enable a user edit their profile



> Please have a get `/profile` route which should call a getProfile method and return a user's profile. Then also have an edit profile route which would call another method and enable a user edit their profile

Okay. This route will standin for the edit-profile.

**Comment by voke1** [comment]
> Please have a get `/profile` route which should call a getProfile method and return a user's profile. Then also have an edit profile route which would call another method and enable a user edit their profile

Done

File: src/middlewares/userMiddlewares.js (hunk 6 above)
Comment (voke1) [reply src/middlewares/userMiddlewares.js:59]: The `checkifUserExistBeforeLogin` checks if user exist during login and gets user email from req.body.
However, the `confirmUserExists` checks if user is valid, during registration. It gets user email  from token sent to the user. As user should not be allowed to update his email during registration.

The two middleware are not exactly the same. Please advise.

File: src/validation/userValidation.js (hunk 11 above)
Comment (voke1) [reply src/validation/userValidation.js]: Done

**Review by uggoprince — APPROVED** [review]
good job @voke1

File: src/middlewares/userMiddlewares.js (hunk 6 above)
Comment (anneKay) [reply src/middlewares/userMiddlewares.js:59]: Okay I have seen this, I would say that you refactor but that might still delay this PR so I will create a chore on the PT board for it instead.

Comment (voke1) [reply src/middlewares/userMiddlewares.js:59]: Okay

File: src/models/userRole.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
+export default (sequelize, DataTypes) => {
```
Comment (anneKay) [inline src/models/userRole.js]: Why do you have this here?

File: public/api-docs/swaggerDoc.json [hunk 13]
```diff
@@ -447,9 +552,37 @@
                 "password": {
                     "type": "string"
                 },
-                "address": {
+                "residencAddress": {
```
Comment (anneKay) [inline public/api-docs/swaggerDoc.json]:
```suggestion
                "residenceAddress": {
```

File: src/models/userRole.js (hunk 12 above)
Comment (voke1) [reply src/models/userRole.js]: This was created as a join table for `many to many relationship` between `user` and `role`

File: src/controllers/UserController.js (hunk 10 above)
Comment (voke1) [reply src/controllers/UserController.js:61]: okay

File: src/models/userRole.js (hunk 12 above)
Comment (anneKay) [reply src/models/userRole.js]: I am wondering what it is doing in this PR

File: public/api-docs/swaggerDoc.json (hunk 13 above)
Comment (anneKay) [reply public/api-docs/swaggerDoc.json]: Please, can you correct this to "residenceAddress" ?
